### ISSUE TITLE
(IntegrityError) duplicate key value violates unique constraint "ix_item_arn"

### ISSUE THREAD

**Issue by riot-gabe** [Netflix/security_monkey#895]
## Please make sure that you have checked the boxes: 

- [x] Review the [Quickstart guide](quickstart.md)
- [x] Search for both open and closed issues regarding the problem you are experiencing
- [x] **For permissions issues** (Access Denied and credential related errors), please refer to the requisite docs before submitting an issue:
[AWS](iam_aws.md), [GCP](iam_gcp.md), [OpenStack](iam_openstack.md), [GitHub](github_setup.md#access-keys-and-permissions)

## Description of issue:
`exceptions` table contains exceptions like `(IntegrityError) duplicate key value violates unique constraint "ix_item_arn"` from source: `scheduler-run-change-reporter`

It looks like `find_changes()` is trying to add some ARNs when they already exist. 

This is a large deployment and we may have somehow screwed up our DB integrity through a separate issue. 

But I would expect the ORM to maybe not think it needs to add a new entry if the ARN already existed.

### Example 1 
Trying to insert `

#### Traceback
 ```
Traceback (most recent call last):
  File "/opt/security_monkey/security_monkey/scheduler.py", line 32, in run_change_reporter
    reporter.run(account, interval)
  File "/opt/security_monkey/security_monkey/reporter.py", line 54, in run
    batch_logic(monitor, monitor.watcher, account, False)
  File "/opt/security_monkey/security_monkey/scheduler.py", line 85, in batch_logic
    audit_items = current_watcher.find_changes(current=items, exception_map=exception_map)
  File "/opt/security_monkey/security_monkey/watcher.py", line 367, in find_changes
    return self.find_changes_batch(current, exception_map)
  File "/opt/security_monkey/security_monkey/watcher.py", line 404, in find_changes_batch
    durable_hash, is_durable)
  File "/opt/security_monkey/security_monkey/datastore_utils.py", line 43, in persist_item
    datastore.db.session.commit()
  File "/opt/security_monkey/venv/local/lib/python2.7/site-packages/sqlalchemy/orm/scoping.py", line 149, in do
    return getattr(self.registry(), name)(*args, **kwargs)
  File "/opt/security_monkey/venv/local/lib/python2.7/site-packages/sqlalchemy/orm/session.py", line 765, in commit
    self.transaction.commit()
  File "/opt/security_monkey/venv/local/lib/python2.7/site-packages/sqlalchemy/orm/session.py", line 370, in commit
    self._prepare_impl()
  File "/opt/security_monkey/venv/local/lib/python2.7/site-packages/sqlalchemy/orm/session.py", line 350, in _prepare_impl
    self.session.flush()
  File "/opt/security_monkey/venv/local/lib/python2.7/site-packages/sqlalchemy/orm/session.py", line 1879, in flush
    self._flush(objects)
  File "/opt/security_monkey/venv/local/lib/python2.7/site-packages/sqlalchemy/orm/session.py", line 1997, in _flush
    transaction.rollback(_capture_exception=True)
  File "/opt/security_monkey/venv/local/lib/python2.7/site-packages/sqlalchemy/util/langhelpers.py", line 57, in __exit__
    compat.reraise(exc_type, exc_value, exc_tb)
  File "/opt/security_monkey/venv/local/lib/python2.7/site-packages/sqlalchemy/orm/session.py", line 1961, in _flush
    flush_context.execute()
  File "/opt/security_monkey/venv/local/lib/python2.7/site-packages/sqlalchemy/orm/unitofwork.py", line 370, in execute
    rec.execute(self)
  File "/opt/security_monkey/venv/local/lib/python2.7/site-packages/sqlalchemy/orm/unitofwork.py", line 523, in execute
    uow
  File "/opt/security_monkey/venv/local/lib/python2.7/site-packages/sqlalchemy/orm/persistence.py", line 64, in save_obj
    mapper, table, insert)
  File "/opt/security_monkey/venv/local/lib/python2.7/site-packages/sqlalchemy/orm/persistence.py", line 594, in _emit_insert_statements
    execute(statement, params)
  File "/opt/security_monkey/venv/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 717, in execute
    return meth(self, multiparams, params)
  File "/opt/security_monkey/venv/local/lib/python2.7/site-packages/sqlalchemy/sql/elements.py", line 317, in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)
  File "/opt/security_monkey/venv/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 814, in _execute_clauseelement
    compiled_sql, distilled_params
  File "/opt/security_monkey/venv/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 927, in _execute_context
    context)
  File "/opt/security_monkey/venv/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1076, in _handle_dbapi_exception
    exc_info
  File "/opt/security_monkey/venv/local/lib/python2.7/site-packages/sqlalchemy/util/compat.py", line 185, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb)
  File "/opt/security_monkey/venv/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 920, in _execute_context
    context)
  File "/opt/security_monkey/venv/local/lib/python2.7/site-packages/sqlalchemy/engine/default.py", line 425, in do_execute
    cursor.execute(statement, parameters)
IntegrityError: (IntegrityError) duplicate key value violates unique constraint "ix_item_arn"
DETAIL:  Key (arn)=(arn:aws:iam::123456789012:role/myrole) already exists.
 'INSERT INTO item (region, name, arn, latest_revision_complete_hash, latest_revision_durable_hash, tech_id, account_id, latest_revision_id) VALUES (%(region)s, %(name)s, %(arn)s, %(latest_revision_complete_hash)s, %(latest_revision_durable_hash)s, %(tech_id)s, %(account_id)s, %(latest_revision_id)s) RETURNING item.id' {'latest_revision_durable_hash': '5ba2d6b8e92a1b5b5676018cd822cb47', 'name': 'myrole', 'latest_revision_complete_hash': '5ba2d6b8e92a1b5b5676018cd822cb47', 'region': 'universal', 'tech_id': 7, 'latest_revision_id': None, 'arn': 'arn:aws:iam::123456789012:role/myrole', 'account_id': 91}

```

### Example 2

Already exists:
```
735284	universal	AWSServiceRoleForOrganizations	7	91	7568596	arn:aws:iam::123456789012:role/aws-service-role/organizations.amazonaws.com/AWSServiceRoleForOrganizations	105cbbd74f8e05830770c0740b9869c0	105cbbd74f8e05830770c0740b9869c0
```
Trying to insert: 
```
{'latest_revision_durable_hash': '105cbbd74f8e05830770c0740b9869c0', 'name': 'AWSServiceRoleForOrganizations', 'latest_revision_complete_hash': '105cbbd74f8e05830770c0740b9869c0', 'region': 'universal', 'tech_id': 7, 'latest_revision_id': None, 'arn': 'arn:aws:iam::123456789012:role/aws-service-role/organizations.amazonaws.com/AWSServiceRoleForOrganizations', 'account_id': 88}
```

#### Traceback
```
Traceback (most recent call last):
  File "/opt/security_monkey/security_monkey/scheduler.py", line 32, in run_change_reporter
    reporter.run(account, interval)
  File "/opt/security_monkey/security_monkey/reporter.py", line 54, in run
    batch_logic(monitor, monitor.watcher, account, False)
  File "/opt/security_monkey/security_monkey/scheduler.py", line 85, in batch_logic
    audit_items = current_watcher.find_changes(current=items, exception_map=exception_map)
  File "/opt/security_monkey/security_monkey/watcher.py", line 367, in find_changes
    return self.find_changes_batch(current, exception_map)
  File "/opt/security_monkey/security_monkey/watcher.py", line 404, in find_changes_batch
    durable_hash, is_durable)
  File "/opt/security_monkey/security_monkey/datastore_utils.py", line 43, in persist_item
    datastore.db.session.commit()
  File "/opt/security_monkey/venv/local/lib/python2.7/site-packages/sqlalchemy/orm/scoping.py", line 149, in do
    return getattr(self.registry(), name)(*args, **kwargs)
  File "/opt/security_monkey/venv/local/lib/python2.7/site-packages/sqlalchemy/orm/session.py", line 765, in commit
    self.transaction.commit()
  File "/opt/security_monkey/venv/local/lib/python2.7/site-packages/sqlalchemy/orm/session.py", line 370, in commit
    self._prepare_impl()
  File "/opt/security_monkey/venv/local/lib/python2.7/site-packages/sqlalchemy/orm/session.py", line 350, in _prepare_impl
    self.session.flush()
  File "/opt/security_monkey/venv/local/lib/python2.7/site-packages/sqlalchemy/orm/session.py", line 1879, in flush
    self._flush(objects)
  File "/opt/security_monkey/venv/local/lib/python2.7/site-packages/sqlalchemy/orm/session.py", line 1997, in _flush
    transaction.rollback(_capture_exception=True)
  File "/opt/security_monkey/venv/local/lib/python2.7/site-packages/sqlalchemy/util/langhelpers.py", line 57, in __exit__
    compat.reraise(exc_type, exc_value, exc_tb)
  File "/opt/security_monkey/venv/local/lib/python2.7/site-packages/sqlalchemy/orm/session.py", line 1961, in _flush
    flush_context.execute()
  File "/opt/security_monkey/venv/local/lib/python2.7/site-packages/sqlalchemy/orm/unitofwork.py", line 370, in execute
    rec.execute(self)
  File "/opt/security_monkey/venv/local/lib/python2.7/site-packages/sqlalchemy/orm/unitofwork.py", line 523, in execute
    uow
  File "/opt/security_monkey/venv/local/lib/python2.7/site-packages/sqlalchemy/orm/persistence.py", line 64, in save_obj
    mapper, table, insert)
  File "/opt/security_monkey/venv/local/lib/python2.7/site-packages/sqlalchemy/orm/persistence.py", line 594, in _emit_insert_statements
    execute(statement, params)
  File "/opt/security_monkey/venv/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 717, in execute
    return meth(self, multiparams, params)
  File "/opt/security_monkey/venv/local/lib/python2.7/site-packages/sqlalchemy/sql/elements.py", line 317, in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)
  File "/opt/security_monkey/venv/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 814, in _execute_clauseelement
    compiled_sql, distilled_params
  File "/opt/security_monkey/venv/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 927, in _execute_context
    context)
  File "/opt/security_monkey/venv/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1076, in _handle_dbapi_exception
    exc_info
  File "/opt/security_monkey/venv/local/lib/python2.7/site-packages/sqlalchemy/util/compat.py", line 185, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb)
  File "/opt/security_monkey/venv/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 920, in _execute_context
    context)
  File "/opt/security_monkey/venv/local/lib/python2.7/site-packages/sqlalchemy/engine/default.py", line 425, in do_execute
    cursor.execute(statement, parameters)
IntegrityError: (IntegrityError) duplicate key value violates unique constraint "ix_item_arn"
DETAIL:  Key (arn)=(arn:aws:iam::123456789012:role/aws-service-role/organizations.amazonaws.com/AWSServiceRoleForOrganizations) already exists.
 'INSERT INTO item (region, name, arn, latest_revision_complete_hash, latest_revision_durable_hash, tech_id, account_id, latest_revision_id) VALUES (%(region)s, %(name)s, %(arn)s, %(latest_revision_complete_hash)s, %(latest_revision_durable_hash)s, %(tech_id)s, %(account_id)s, %(latest_revision_id)s) RETURNING item.id' {'latest_revision_durable_hash': '105cbbd74f8e05830770c0740b9869c0', 'name': 'AWSServiceRoleForOrganizations', 'latest_revision_complete_hash': '105cbbd74f8e05830770c0740b9869c0', 'region': 'universal', 'tech_id': 7, 'latest_revision_id': None, 'arn': 'arn:aws:iam::123456789012:role/aws-service-role/organizations.amazonaws.com/AWSServiceRoleForOrganizations', 'account_id': 88}

```

**Comment by riot-gabe** [comment]
Oh, hey. I just noticed what we are trying to insert has a different account ID than what exists. 
Upon further investigation, we have two duplicate accounts in our account table. Not sure how that happened. 

I'm going to clean that up in my environment. 

Maybe we can turn this into some sort of "don't allow duplicate account identifier" assertion when adding accounts. 

It looks like there /might/ be a case where we accept identifier of `None`? We have a line for `name: Cloudfront logs` that I'm not sure if we added ourselves or not with identifier `None`

If that was another weird-thing just for us, perhaps we might consider a `UNIQUE` constraint. But there may be other considerations. 

I'm going to close this for now. Maybe it will be useful for another person someday.
